ESPS/general/src/setrange/setrange.1
.\" Copyright (c) 1986-1990 Entropic Speech, Inc.
.\" Copyright (c) 1996 Entropic Research Laboratory, Inc.  All rights reserved.
.\" @(#)setrange.1	1.9 3/31/97 ESI/ERL
.ds ]W (c) 1991 Entropic Research Laboratory, Inc.
.TH SETRANGE 1\-ESPS 3/31/97
.SH NAME
setrange - put range in records or seconds in Common; allow conversion from seconds
.SH SYNOPSIS
.B setrange
[
.BI \-x " debug_level"
] [
.BI \-P " param_file"
] [
.BI \-{prs} " range"
] [
.B \-z
] 
.I " esps_file"
.SH DESCRIPTION
.PP
.I Setrange
takes a range of time (specified in seconds), converts it to a range
of records in the ESPS FEA file \fIesps_file\fP, and leaves
the sample range in ESPS Common.  Note that \fIsetrange\fP
takes into consideration the value of the
\fIstart_time\fP generic, if it exists.  (That is, the times 
specified by \fB-s\fP are treated as absolute times, not as 
times relative to the start of the file.)   \fISetrange\fP also 
takes a range specified in records, converts it to a range in 
seconds, and leaves the time range in ESPS Common.  Actually, 
regardless of whether a time or record range is given, \fIsetrange\fP
writes Common with range values in both measurement units.  
.PP
In computing the range in records, \fIsetrange\fP uses the generic
\fIrecord_freq\fP (number of records per seconds).  This is the
sampling rate for FEA_SD files.  If \fIrecord_freq\fP does not exist
or is less than or equal to 0, \fIsetrange\fP exits with an error
message.
.PP
Unless the \fB-z\fP option is used, \fIsetrange\fP prints a message
giving the range in points that is stored in ESPS Common.
.PP
\fISetrange\fP will work properly with old-style sampled data (SD)
files.  
.PP
\fISetrange\fP is useful when run before other ESPS programs that
process FEA files but that do not permit a range specification in time
units.  It is also useful in scripts when one needs to convert in
either direction between range specs in records or time.  Future
versions of ESPS will provide a uniform means of specifying ranges as
time or samples.  \fISetrange\fP is provided as interim support.
.SH OPTIONS
.PP
The following options are supported:
.TP
.BI \-x " debug"
Only debug level 1 is defined in this version.   This causes several
messages to be printed as internal processing proceeds.   The default
is level 0, which causes no debug output.
.TP	 
.BI \-P " param_file"
uses the parameter file 
.I param_file
rather than the default, which is
.I params.
.TP
.BI \-s " first" : "last"
.TP
.BI \-s " first" :+ "incr"
Specifies in units of time (seconds) the range of sampled data in
\fIesps_file\fP.  In the first form, a pair of unsigned integers
specifies the range.  If
.IR last " = " first " + " incr,
the second form (with the plus sign) specifies the same range as the
first two forms.  If 
.I first
is omitted, the default value of 0 is used.  If 
.I last
is omitted, then the end of the file is assumed. 
The beginning of \fIesps_file\fP is treated as 0 seconds.  
.TP
.BI \-{p} " first" : last "\fR [1:(last in file)]"
.TP
.BI \-{p} " first" :+ incr
.TP
.BI \-{r} " first" : last "\fR [1:(last in file)]"
.TP
.BI \-{r} " first" :+ incr
Specifies in units of points the range of sampled data in the
\fIesps_file\fP.  In the first form, a pair of unsigned integers gives
the first and last points of the range, counting from 1 for the first
point in the file.  If
.I first
is omitted, 1 is used.  If 
.I last 
is omitted, the range extends to the end of the file.  The second form
is equivalent to the first with
.I "last = first + incr".
This option should not be specified if
.B \-s
is specified.  If neither option is specified, the range is determined
by the parameters
.I start_s
and
.I nan_s
as read from the parameter file.  If either parameter is missing from
the parameter file, it is determined by default.  Note that \fB-p\fP
and \fB-r\fP are synonyms. 
.TP
.B \-z
Suppress output message giving range of sampled data points.  
.SH ESPS PARAMETERS
.PP
The parameter file is not required to be present, as there are 
default parameter values that apply.  If the parameter file 
does exist, the following parameters are read:
.TP
.I "start_s - float"
.IP
The starting point to be processes (in seconds).  A value of 0 denotes
the start of the file.  This parameter is read only if the no range
option is used.  If it is not in the parameter file, the default value
of 0 is used.
.TP
.I "nan_s - float"
.IP
The total number of seconds in the range.  If 
.I nan_s
is 0, the whole file is processed.  This parameter is read only if the no range
option is used.  If it is not in the parameter file, the default value
of 0 is used.
.SH ESPS COMMON
.PP
ESPS Common is not read.  
.PP
The following items are written into the ESPS Common file:
.IP
.I "start - integer"
.IP
The starting sample number in \fIesps_file\fP.
.IP
.I "nan - integer"
.IP
The number of samples in the selected range.
.IP
.I "start_s - float"
.IP
The starting time (seconds) in \fIesps_file\fP.
.IP
.I "end_s - float"
.IP
The ending time (seconds) in the selected range.
.IP
.I "nan_s - float"
.IP
The totoal time (seconds) in the selected range.
.IP
.I "prog - string"
.IP
This is the name of the program (\fIsetrange\fP in this case).
.I "filename - string"
.IP
The name of \fIesps_file\fP. 
.SH ESPS HEADERS
.PP
For FEA files, \fIsetrange\fP makes use of the the generic header
items start_time and record_freq.  If the input file is an old
style SD, the header is converted automatically to a FEA_SD header 
before proceeding (i.e., the generics don't need to be there in the 
SD header).  
.SH DIAGNOSTICS
.PP
.I Setrange
informs the user and exits if \fIesps_file\fP does not exist, or is not an
ESPS sampled data file.
.PP
If the starting point requested is greater than the last point in
\fIesps_file\fP, then a message is printed and the program exits.  If
the ending point requested is greater than the last point in
\fIesps_file\fP, it is reset to the last point and a warning is
printed.  continues.
.SH FILES
.SH BUGS
.SH EXPECTED CHANGES
.PP
.SH AUTHOR
.PP
Manual page and program by John Shore.  




ESPS/general/src/setrange/setrange.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1988-1990  Entropic Speech, Inc. 
 *    "Copyright (c) 1990-1996  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  John Shore
 * Checked by:
 * Revised by: Ken Nelson, John Shore
 *
 * Brief description:
 *  This program puts in common a sampled data range (in points) given
 *  a sampled data range in seconds.  
 */

#ifndef lint
static char *sccs_id = "@(#)setrange.c	1.14	06 Oct 1999 ESI";
#endif

int debug_level = 0;

#include <stdio.h>
#include <esps/esps.h>
#include <esps/unix.h>
#include <esps/fea.h>

#define REQUIRE(test,text) {if (!(test)) {(void) fprintf(stderr, \
"%s: %s - exiting\n", ProgName, text); exit(1);}}

#define SYNTAX \
USAGE("setrange [-{prs} range] [-x debug_level] [-z] sdfile") ;

struct header  *sdtofea();
void	frange_switch();
char	*ProgName = "setrange";

main(argc, argv)
    int  argc;
    char **argv;
{
    extern int
	    optind;		/* for use of getopt() */
    extern char
	    *optarg;		/* for use of getopt() */
    int	    ch;			/* command-line option letter */
    char    *srange;		/* range string from -s */
    int	    sflag = 0;		/* -s option specified? */
    int	    pflag = 0;		/* -p option specified? */
    int	    zflag = 0;		/* -z option (suppress) specified? */
    double  start_s;		/* start second in the input file */
    double  nan_s;		/* total time of range */
    double  end_s;		/* end of range in seconds*/
    double  start_time = 0;     /* start_time generic */
    long    start_p;		/* starting sample */
    long    end_p;		/* last sample */
    long    nrec;		/* number of records in file */
    long    nan;		/* total number of samples to analyze */
    float   isf = 0;		/* input file sampling frequency */
    char    *iname;		/* input file name */
    FILE    *ifile;		/* input stream */
    char    *param_name = NULL;
				/* parameter file name */
    struct header
	    *hd;		/* input file header */
/* Parse command-line options. */

    while ((ch = getopt(argc, argv, "P:p:r:s:x:z")) != EOF)
        switch (ch)
	{
	case 'P':
	    param_name = optarg;
	    break;
	case 's':
	    srange = optarg;
	    sflag++;
	    break;
	case 'p':
	case 'r':
	    srange = optarg;
	    pflag++;
	    break;
	case 'z':
	    zflag++;
	    break;
	case 'x':
	    debug_level = atoi(optarg);
	    break;
	default:
	    SYNTAX
	    break;
	}

    if (argc - optind > 1) {
	Fprintf(stderr,
	    "%s: too many file names specified.\n", ProgName);
	SYNTAX
    }

    if (argc - optind < 1) {
	Fprintf(stderr,
	    "%s: must specify a FEA file name.\n", ProgName);
	SYNTAX
    }

    /*we open the file; note, we don't care what file type it is*/

    iname = argv[optind++];

    TRYOPEN (ProgName, iname, "r", ifile);

    if (!(hd = read_header (ifile)))
	NOTSPS (ProgName, iname);

    if (debug_level)
	Fprintf(stderr, "Input file: %s\n", iname);

    /*convert to FEA_SD if FT_SD*/

    if (hd->common.type == FT_SD)
	hd = sdtofea(hd);

    isf = get_genhd_val("record_freq", hd, 0.0);

    if (debug_level)
	Fprintf(stderr, "sampling rate = %g\n", isf);

    REQUIRE(isf > 0,
	    "generic record_freq (sample fequency) is <= 0 or not found");

    nrec = hd->common.ndrec;
    if (nrec == -1)
    {
	struct fea_data	*rec;

	REQUIRE(hd->common.type == FT_FEA, "non-FEA file can't be on a pipe");

	/* Count records by reading */

	rec = allo_fea_rec(hd);
	for (nrec = 0; get_fea_rec(rec, hd, ifile) != EOF; nrec++)
	{ }

	free_fea_rec(rec);
    }

    (void) read_params(param_name, SC_NOCOMMON, iname);

/*
 * Get start_time from input file - defaults to zero if not there
 */
    start_time = get_genhd_val("start_time", hd, (double)0);

/*
 * Process range option
 */
    if (pflag) {
	start_p = 1;
	end_p = LONG_MAX;
	lrange_switch(srange, &start_p, &end_p);
	REQUIRE(end_p >= start_p, "end can't be before the start");
    	if (end_p > nrec) {
	    end_p = nrec;
	    Fprintf(stderr,
		"%s: Warning - range extends beyond file; truncated\n",
		ProgName);
	}
	start_s = start_time + (start_p - 1)/isf;
	end_s = start_time + (end_p - 1)/isf;
	nan_s = end_s - start_s;
	nan = end_p - start_p + 1;

    }

    if (sflag) {
	start_s = 0;
	end_s = DBL_MAX;
	frange_switch(srange, &start_s, &end_s);
    }
    if (!sflag && !pflag) {
	start_s =
	    (symtype("start_s") != ST_UNDEF)
	    ? getsym_d("start_s")
	    : 0;

	if (symtype("nan_s") != ST_UNDEF && getsym_d("nan_s") != 0)
	    end_s = start_s + getsym_d("nan_s");
	else
	    end_s = DBL_MAX;

	sflag = 1;
    }
    if (sflag) {
	REQUIRE(end_s >= start_s, "end can't be before the start");

	start_p = (start_s - start_time)*isf + 1;

	if (start_p < 1) {
	    start_p = 1;
	    Fprintf(stderr,
		    "%s: Warning - range starts before file; truncated\n",
		    ProgName);
	    start_s = start_time;

	}

	if (end_s == DBL_MAX)
	    end_p = nrec;
	else
	    end_p = end_s * isf + 1;

	if (end_p > nrec) {
	    end_p = nrec;
	    end_s = start_time + (end_p - 1)/isf;
	    Fprintf(stderr,
		    "%s: Warning - range extends beyond file; truncated\n",
		    ProgName);
	}

	nan = end_p - start_p + 1;
	nan_s = end_s - start_s;
    }

    if (debug_level) {
	Fprintf(stderr,
		"%s: start_s = %g, end_s = %g\n", ProgName, start_s, end_s);
	Fprintf(stderr,
		"%s: start_p = %ld, end_p = %ld, nan = %ld\n",
		ProgName, start_p, end_p, nan);
    }

    if (!zflag) {
	Fprintf(stdout,
		"%s: starting record = %ld, number of records = %ld\n",
		ProgName, start_p, nan);
    }
/*
 * put info in ESPS common
 */
    if (end_s == DBL_MAX)
	end_s = FLT_MAX;
    if (nan_s == DBL_MAX)
	nan_s = FLT_MAX;

    (void) putsym_s("filename", iname);
    (void) putsym_s("prog", ProgName);
    (void) putsym_i("start", (int) start_p);
    (void) putsym_i("nan", (int) nan);
    (void) putsym_f("start_s", (float) start_s);
    (void) putsym_f("end_s", (float) end_s);
    (void) putsym_f("nan_s", (float) nan_s);

    exit(0);
    /*NOTREACHED*/
}



